tensorflow/core/graph/mkl_layout_pass.cc

```diff
@@ -2516,6 +2915,29 @@ bool MklLayoutRewritePass::RunPass(std::unique_ptr<Graph>* g) {
 
   DumpGraph("After running MklLayoutRewritePass(NodeMerge)", &**g);
 
+  order.clear();
+  GetReversePostOrder(**g, &order);  // This will give us topological sort.
+  for (Node* n : order) {
```

Comment:
This pass happening after node merging means it could miss an opportunity for a long fusion. For example, for a graph A->B->C->D->E and merge rules {A->B to AB, D->E to DE} and fusion rule {A->B->C->D->E to ABCDE}, the result will be AB->C->DE instead of ABCDE.

If you plan to do a long fusion chain only through merge rules, i.e., {A->B to AB, AB->C to ABC, ...} and only use fusion rules that doesn't contain any patterns already occurred in merge rules, please add a comment clarifying this.